### ISSUE TITLE
Roadmap to 1.0

### ISSUE THREAD

**Issue by BusterNeece** [AzuraCast/AzuraCast#4460]
Over the years, many folks have asked us what it would take for us to tag a 1.0 release version of AzuraCast. After all, thousands of stations run every day on AzuraCast already without any issue, and we have a stable enough infrastructure that we're able to tag "Stable" versions of our software every few months.

Historically, we've (well, _I've_) been hesitant to ever tag a 1.0 release version for the same reason that a lot of projects are in a sort of "perma-beta" state: because people have different expectations for something they see as under active development versus something they see as "released". We've been faced with very upset users in the past and we don't want to encourage that behavior via our own actions.

So, here's the bottom line: it will never be possible for us to write completely bug-free software or to eliminate every issue that exists prior to a 1.0 launch, but we will manage the process in such a way that everyone has plenty of opportunities to report bugs to us before the final tagged release version, and it's our hope that the resulting product will be one that is reliable, powerful and that folks can count on for production radio stations the way many stations already do.

With that in mind, there are a few things that I think are necessary for AzuraCast to have in place before we can tag a 1.0 release. We've made a lot of headway on these items already, but I'd want them to be done before we tag a 1.0 version:

- [x] **Implement Liquidsoap 2.0**: The next major release of Liquidsoap is already in its testing phases, and we already have branches set up (#4402 and azuracast/docker-azuracast-radio#12) to test how it integrates with AzuraCast; we will want 2.0 to itself have a stable tagged release, and for our integration of it to be well-tested and merged in our main branch.

- [x] **HLS Support**: We have a separate issue for our proposed implementation of this (#3685). We expect this will become somewhat easier to implement with the adoption of Liquidsoap 2.0.

- [ ] **Grouped/Nested Playlists**: One of the most widely-requested features that we don't currently support is the idea of "nested" playlists; one overarching playlist that contains other ones in a specified playback order. For example, one parent playlist could say "Play X songs from playlist 1, then Y songs from playlist 2, then Y from playlist 3". Although our code doesn't currently support this, implementing it is, at least in my opinion, imperative for a 1.0 release.

- [x] **"Vue-tify" Everything**: Already, as we build new functionality into AzuraCast, we're slowly moving components over to Vue pages that call underlying APIs to accomplish their work. There are still areas that haven't been fully Vue-tified, though:]
   - [x] Dashboard
   - [x] Administration
   - [x] API Keys
   - [x] Backup Settings / Run Backup
   - [x] Custom Fields
   - [x] Install GeoLite
   - [x] Install SHOUTcast 2
   - [x] User Profile + 2-Factor Setup
   - [x] User Registration
   - [x] Permissions/Roles
   - [x] System Settings
   - [x] Custom Branding
   - [x] Clone Station
   - [x] Audit Log
   - [x] Administer Users
   - [x] Station: Playlist Automation
   - [x] Station: SFTP Users
   - [x] Station: Profile View
   - [x] Station: Profile Edit
   - [x] Station: Mount Points
   - [x] Station: Remote Relays
   - [x] Station: Web Hooks
   - [x] Station: WebDJ
   - [x] Station: SoundExchange Report
   - [x] Station: Statistics Overview Report
   - [x] Station: Live Listeners Report
   - [x] Station: Song Listener Impact Report
   - [x] Station: Playback Timeline Report

- [x] **Evaluate Media Processing**: In recent versions, we've implemented code that separates the media processing process from the rest of the system. The media processing code doesn't need to be PHP at that point, and may even benefit heavily by _not_ being PHP; we should evaluate our current processing performance metrics and whether spinning that program off into a C#, Golang, etc. program would be appropriate.

- [ ] **Full Localization in Selected Languages**: We rather frequently change strings, but for the 1.0 release process, we would intend to have a "string freeze" early in the development cycle to allow for our friends who help with translation and localization to fully translate the entire application. Currently, our ten most localized languages are the ones below, which we would expect to be at 100% translated strings before any 1.0 release:
   - [ ] German (de-DE)
   - [ ] French (fr-FR)
   - [ ] Dutch (nl-NL)
   - [ ] Polish (pl-PL)
   - [ ] Spanish (es-ES)
   - [ ] Russian (ru-RU)
   - [ ] Turkish (tr-TR)
   - [ ] Portuguese (pt-PT)/Brazilian Portuguese (pt-BR)
   - [ ] Greek (el-EL)
   - [ ] Simplified Chinese (zh-CN)

---

I would love to hear folks' thoughts on this. I know everyone has a particular feature or two that they're especially interested in, and I don't want this thread to just become folks listing out the specific features they want for their own station, but rather a big-picture discussion about what we should prioritize as we target a 1.0 release that would do the most good for the greatest number of stations.

**Comment by ogrenci01** [comment]
Hello;

I love Azuracast. I've been using it for over 2 years with no problems. I like the Ansible version more than Docker. Azuracast should now exit beta. I agree with this thought.

I translated the entire Azuracast into Turkish.

I have one more thought. I hope that will be useful.

Is it possible to make a firewall with a graphical interface that works with Azuracast? I don't know how Docker works on this. I am actively using CSF for Ansible. I have tips for some safety precautions. I'm very sorry, but the reality of DDoS/Botnet still exists today. Depending on this it may be necessary to block some bad traffic. This is forcing the server owners. For example, we can add "ratelimit" rules that will not cause problems for users.

Hope this project will continue for many years. That's all I want to say for now. Endless thanks to the team.

**Comment by idolaru** [comment]
You can have my support for native Romanian, Chinese Simplified and Indonesian. Feel free to contact me.

**Comment by finnie2006** [comment]
I can help with Dutch, Feel free to contact me.

**Comment by BenFromXonos** [comment]
@ogrenci01 

> Is it possible to make a firewall with a graphical interface that works with Azuracast? I don't know how Docker works on this. I am actively using CSF for Ansible. I have tips for some safety precautions. I'm very sorry, but the reality of DDoS/Botnet still exists today. Depending on this it may be necessary to block some bad traffic. This is forcing the server owners. For example, we can add "ratelimit" rules that will not cause problems for users.

I currently use a Docker install which is obviously a bit different, but I personally block the radio ports using UFW, whitelist the Nginx proxy inside AzuraCast to only be accessible through a "proxy" server and I run rate-limiting and other proprietary DDoS protection techniques on that proxy.

This helps a LOT with the threat of botting and DDoS attacks.

I feel like something like this would be cool to add to AzuraCast, BUT I think it would be slightly out of the project's scope.

**Comment by GeertSurkijn** [comment]
Hello, I am looking forward to seeing the Grouped/Nested Playlists feature. This will be a very nice improvement in automating your radio. One important legal issue may not be forgotten for this. Pretty sure this will be a challenge and most likely is already discussed between the developers. That is, that such a grouped/nested playlist must also have the "Avoid Duplicate Artists/Titles" option and handle with care.

My personal opinion on this matter is that it should be removed from the playlist setting, but must work station wide. Of course, it can only take care of this on the queued tracks for the random types. The live and sequenced playlist is the responsibility of the DJ's.

Further, in some countries, the rules on streaming radio are a bit more restrictive than the current autodj setting is offering.
4 hours between the same track/song
2 hours between tracks from the same album
1 hour between tracks from the same artist

**Comment by realhardstyle** [comment]
Hi,

As recommended I'm posting my use case of the Grouped/Nested Playlists feature to see if it can be built-in.
Let me explain it a little bit. Our DJs regularly upload a mix via the website. The mix name always contains the date of the broadcast.
This mix is uploaded to a folder with the DJ name in it. A (Powershell) script runs to set the idv3 tags and checks if there is more than 1 mix in the folder. If there is then he moves the oldest mix to an archive folder specific to that DJ. 

Besides that script, another script runs to create SymLinks. These symlinks are created based on the date in the filename. So if a mix has the date of today, The symlink points to the active folder of the DJ. When there is no new set for today (so no filename with the date of today) then the symlink is created to the Archive folder of the DJ.

In this way, the DJ is flexible when he uploads a new mix and one will be played from the archive when he does not upload.
I hope this is making sense.

Would be awesome if this could be done within AzuraCast playlists. We currently have not yet moved to AzuraCast because we haven't found a solution for it. Also, StereoTool is required for us but isn't integrated and we are having trouble getting it working.

Kind Regards,

**Comment by Pretmaker** [comment]
It might be an unpopular opinion, but I feel that Azuracast itself is a long long long way from becoming a 1.0 release still.

I personally dislike docker allot. Reason being is that it's a hassle to change (configuration) files permanently, only supported changes are in the SQL database.
Liquidsoap implementation lacks allot of advanced features, like better in-line processing (I have figured it out how to use it after spending 40 hours on trial and error on stuff and got professional processing up and running, not the baby liquidsoap processing but comparable processing to digital radio stations with 10,000s of dollar funding). This in all honesty took way too much time as it's an "advanced" option and almost completely breaks the current liquidsoap install.

So imho you'd either have to better implement those options also for the less technical users (as the liquidsoap clear script isn't clear enough and needs more clarification), or you should remove the option all together and make it a minimum viable product.

As for the options, you have no option currently on what you publish currently for the "now playing". So it's nice to see that some stations who act big only have like 5-50 listeners. I doubt you'd want to be looked down on, thus I advice to make this configurable, or remove the nowplaying API access to make it a minimum viable product.

Connected clients (shoutcast) don't receive the relayed stats of the current listener amount. Most DJ software however does support this. It would be advisable to see if this can be added for the 1.0 release. Not sure if this is a liquidsoap limiting ting, but it might be. Not done research on it yet, Although it is possible to request the listener stats every 3 seconds for example via the nowplaying API which just publishes/adds up the icecast listener count.

Als in regards to the icecast configuration, also in the docker container installation, the stats will be publicly available. It might be nice to have an option to disable that also, unless you want people to know you've got like 5 listeners per relay XD.

Then there's the current GDPR issues where people could technically request any and all information from a station. This can include statistics that are in the database. How the hell am I going to extract and verify and delete that info/how do I know that it doesn't have the records? As when I look at the stats I can still see countries were people listened from and I see current IP clients and application info. So there is information logged and I don't have proof that I don't have it logged. Note I'm not a lawyer, would help if an actualy lawyer could verify things like this.
For this reason I've got recordholding set to the shortest amount possible, and instead rely on requesting the API for information and saving it to my websites database.

I concure with some of the previous mentions of things to add, like an application firewall. Same with some more security in front of icecast for example.
Icecast can be brute-forced if you don't use fail-2-ban for example for the admin interface which is my soluation in the ansible installation with 4 relays in total to spread the listeners around.

Small thing to note with the liquidsoap, the songname that gets pushed to the ice-cast servers could use an API setting. I don't like it that liquidsoap uses the metadata from the presenters who connect as they all end up with different non-standard metadata. Thus resulting in song statistics being unreliable, and songnames to become duplicates. 
I've solved this myself by letting my webserver make an API call to the icecast servers to change the songname to a standard that my webserver figures out with a Spotify lookup/integration. Works for like 98% of pretty much all of the regular songs presenters play.

I hope some of my points are clear enough of what I think needs some extra love from the development perspective, perhaps even a redesign or some sorts to give more control to the operators. 

Talking about operators, I feel that the interface is a bit lack-luster. It is an unexplainable feeling as I know for example that some other products are WAY worse UX orientated. But I do feel that there can be some gains.

On the field of the translations, I feel that some translations are a bit "stupid".
For example:
On the profile page, On the Air (in Dutch language) shows as On the Air...
Same on the profile page, The 2 phrases of enable/disble... It's translated to the literal translation of "Schakel uit" and "Schakel in". The main question here is, how was the orriginal line meant? If I would translate from Dutch to English I would get: Turn Off, Turn On... I'm like what the hell am I "turning off or on"? Am I not just "enabling/disabling" an option? At which point the actual verb usage for execution would be better (uitschakelen, inschakelen). This is also 1 word instead of 2 words which would come back to the UX allowing for it to become actual buttons.
Same with some missing translations, when I click Mount points (still in Dutch interface, no idea why it's called Mount points as it's not a good "Dutch" translation), there is an "information message" in English, and other things are in Dutch once again. 
This makes it feel to me that overal the translation tried to do a good job with translating the basics, but lots of things are perhaps still hard coded? And wanna know the worst part of it? 
Speaking of UX... It took me way too much time to find where I can change the display language. 

I might edit later to agree with others or post again. But I feel that there's still allot of work to be done before a 1.0 release would be a succes. Lots of polishing needed still

**Comment by SlvrEagle23** [comment]
> It might be an unpopular opinion, but I feel that Azuracast itself is a long long long way from becoming a 1.0 release still.

I appreciate the feedback, but I want to note that your account, per my check, shows zero GitHub issues, zero Pull Requests, zero Discussions, and zero donations.

We don't take marching orders from people who aren't willing to walk alongside us. That's not how open-source works.

> I personally dislike docker allot. Reason being is that it's a hassle to change (configuration) files permanently, only supported changes are in the SQL database. Liquidsoap implementation lacks allot of advanced features, like better in-line processing (I have figured it out how to use it after spending 40 hours on trial and error on stuff and got professional processing up and running, not the baby liquidsoap processing but comparable processing to digital radio stations with 10,000s of dollar funding). This in all honesty took way too much time as it's an "advanced" option and almost completely breaks the current liquidsoap install.

Ansible is available for power users, and still works to this day. We don't advertise it, and yet we maintain it specifically for people who, like this, dislike Docker for what it is.

Docker is what we recommend for the vast majority simply because it allows us to build and configure one ideal image with everything set up exactly as it should be, then deploy that _exact_ image out to thousands of users. It's literally the only way a team of 2 devs and a community manager could ever oversee an operation of this scale. Trust us, we tried it the other way first.

> So imho you'd either have to better implement those options also for the less technical users (as the liquidsoap clear script isn't clear enough and needs more clarification), or you should remove the option all together and make it a minimum viable product.

AzuraCast was a 'minimum viable product' in 2016. Today it powers countless stations. Just because it doesn't work well for your needs doesn't mean it doesn't work well.

> As for the options, you have no option currently on what you publish currently for the "now playing". So it's nice to see that some stations who act big only have like 5-50 listeners. I doubt you'd want to be looked down on, thus I advice to make this configurable, or remove the nowplaying API access to make it a minimum viable product.

You can definitely stop using the term "minimum viable product" at this point, as it's clear you're describing what _you, specifically_, want out of a radio management suite.

> Connected clients (shoutcast) don't receive the relayed stats of the current listener amount. Most DJ software however does support this. It would be advisable to see if this can be added for the 1.0 release. Not sure if this is a liquidsoap limiting ting, but it might be. Not done research on it yet, Although it is possible to request the listener stats every 3 seconds for example via the nowplaying API which just publishes/adds up the icecast listener count.

This _is_ a Liquidsoap thing. We could've told you this if you had ever asked us.

> Als in regards to the icecast configuration, also in the docker container installation, the stats will be publicly available. It might be nice to have an option to disable that also, unless you want people to know you've got like 5 listeners per relay XD.

Hear me out, but some stations broadcast for the love of music, or their desire to show their specific curated playlist off to the world. I've known many stations to get overwhelmingly obsessed with listenership numbers, and the obsession becomes an unhealthy one quickly.

> Then there's the current GDPR issues where people could technically request any and all information from a station. This can include statistics that are in the database. How the hell am I going to extract and verify and delete that info/how do I know that it doesn't have the records? As when I look at the stats I can still see countries were people listened from and I see current IP clients and application info. So there is information logged and I don't have proof that I don't have it logged. Note I'm not a lawyer, would help if an actualy lawyer could verify things like this. For this reason I've got recordholding set to the shortest amount possible, and instead rely on requesting the API for information and saving it to my websites database.

We specifically limit our analytics collection to help improve our GDPR compliance. You can change those settings at any point from the System Settings panel. As soon as you change the settings to be more restrictive, within a few minutes we purge the system of any records that don't comply with that level of detail.

> Talking about operators, I feel that the interface is a bit lack-luster. It is an unexplainable feeling as I know for example that some other products are WAY worse UX orientated. But I do feel that there can be some gains.

If you cannot explain it, we cannot fix it.

> On the field of the translations, I feel that some translations are a bit "stupid". For example: On the profile page, On the Air (in Dutch language) shows as On the Air... Same on the profile page, The 2 phrases of enable/disble... It's translated to the literal translation of "Schakel uit" and "Schakel in". The main question here is, how was the orriginal line meant? If I would translate from Dutch to English I would get: Turn Off, Turn On... I'm like what the hell am I "turning off or on"? Am I not just "enabling/disabling" an option? At which point the actual verb usage for execution would be better (uitschakelen, inschakelen). This is also 1 word instead of 2 words which would come back to the UX allowing for it to become actual buttons. Same with some missing translations, when I click Mount points (still in Dutch interface, no idea why it's called Mount points as it's not a good "Dutch" translation), there is an "information message" in English, and other things are in Dutch once again. This makes it feel to me that overal the translation tried to do a good job with translating the basics, but lots of things are perhaps still hard coded? And wanna know the worst part of it? Speaking of UX... It took me way too much time to find where I can change the display language.

Translations are managed by our very generous community submitting them via CrowdIn at https://translate.azuracast.com/. If you find translations in need of correction, please submit them there.

> I might edit later to agree with others or post again. But I feel that there's still allot of work to be done before a 1.0 release would be a succes. Lots of polishing needed still

Great. We agree. Start contributing.

**Comment by SlvrEagle23** [comment]
@gusaus the Statistics Overhaul is still on our to-do list, though is a lower priority and isn't something we would necessarily include in the 1.0 roadmap.

If we get some help with the Drag and Drop scheduler, then we can work on that too, but I wouldn't start work on it myself as I don't have enough experience on implementing that kind of thing myself.

We tried to use the OpenCollective sub-project system for the statistics overhaul, and it worked fine, but it's not a bounty system at all.

**Comment by gusaus** [comment]
@SlvrEagle23 I appreciate the clarity. 

I've seen Drag and Drop scheduler referenced in discussions around features like https://github.com/AzuraCast/AzuraCast/discussions/5407 but haven't seen an RFP or Feature Bounty (similar to https://github.com/AzuraCast/AzuraCast/issues/5242). It's been a while since we discussed your specific needs to move that forward. Are there certain tasks you'd need help with (including non-development tasks like requirements, docs, and fundraising)? Would it make sense to create a separate issue or discussion about this specific feature as it's out of scope for 1.0? 

I'll followup in Open Collective's Slack about how OpenProducer's pilot 'should' provide a way to direct resources towards the development of 'commercial' features as well as this 1.0 release. 

Really appreciate all the work you and the team have put into the project!

**Comment by fnfzlznddzn** [comment]
AzuuraCast is making great progress every day, it's awesome to see. Thank you so much for your continued hard work on this software!

1. In my opinion, the **Drag & Drop Scheduler** is one of the most important features to make AzuraCast easy and practical to use on a daily basis for the radio manager. Currently, changing when playlists are scheduled is quite painful and clunky, in particular when you have a lot of them.

2. I agree that the Nested Playlist idea is also a priority, this will be useful.

3. Having an "autoplaylist generator" based on file metadata would be pretty great too. For example, you specify some tags that must be present (e.g. genre = [hip hop OR rap] AND conscious / date = AFTER 1985 AND BEFORE 2001) or absent (e.g. genre =/= instrumental, gangsta). I currently do this using foobar2000 and export the files into folders with hardlinks (to avoid duplication of tracks present in multiple playlists) which I then import in AzuraCast, but this is definitely not a straightforward solution for most people. I believe having this option natively within AzuraCast would make the radio managing aspect significantly more powerful and interesting.

As for the listener point of view, I think Voting on Songs would be one of the main priorities. It's definitely one of the things people mention to me first.

Thanks for reading!

**Comment by fidegr** [comment]
Hello, a functionality that I think is essential for the proper functioning of a station is the playlists with the option to interrupt so they can respect the assigned priority. For example, only interrupt other playlists with lower weight. If a list with the same or higher weight is playing, wait for it to finish. This would allow meeting schedules without interrupting programs, but without interrupting music playlists.

**Comment by Moonbase59** [comment]
Here’s my ideas, sorted by priority (top=highest):

- bugfixes for handling larger, mounted-in collections (~150k here) (see https://github.com/AzuraCast/AzuraCast/issues/6088)
- Overhaul of the tag reader/writer, using internal names and mappings for different file formats, much like MusicBrainz Picard does it. (see https://github.com/AzuraCast/AzuraCast/issues/6092, https://github.com/AzuraCast/AzuraCast/issues/6093)
- Fast search with complex searches (like Amarok, Clementine or using Meilisearch), to build better playlists within AzuraCast. (see https://features.azuracast.com/suggestions/364412/better-search-for-building-better-playlists).
- Station clockwheel building, also needs grouped/nested playlist feature.
- Drag-n-drop Scheduler.
- Customizable player and schedule widgets to include in station’s homepage. (Done, I guess)
- Good API docs with examples (probably already exist) so dynamic updates for different information can be made easily on a station’s homepage. And CORS, of course.

Thanks & keep it rollin’!

**Comment by BusterNeece** [comment]
Closing this issue as we're no longer really referencing it to target a 1.0 release. Much of the work here has been completed, and our needs for what should be in a 1.0 release are always evolving.
